### ISSUE TITLE
Doc: ctc_beam_search_decoder

### ISSUE THREAD

**Issue by githubharald** [tensorflow/tensorflow#21051]
The [TF documentation ](https://www.tensorflow.org/api_docs/python/tf/nn/ctc_beam_search_decoder) says: _"Note The ctc_greedy_decoder is a special case of the ctc_beam_search_decoder with top_paths=1 and beam_width=1 (but that decoder is faster for this special case)."_

This implies that the result of the greedy and beam decoder is equal if beam width is set to 1.

However, this is not the case. Even with beam width set to 1, beam search uses more information by distinguishing paths ending with blank and non-blank.
Therefore, it is possible to create inputs from which beam search can extract enough information to produce the correct result, while best path decoding fails.

Here is such an input matrix with 3 time-steps. It contains a label '0' and the blank '-'. The best path "0-0" is marked by a red line:
![matrix_best](https://user-images.githubusercontent.com/15148095/43075266-908f4740-8e80-11e8-85a5-91e9ffd08015.png)

 

And here is the code which produces different results, depending on the decoder:

```
import tensorflow as tf
import numpy as np


batchSize=1
numClasses=2
numTimesteps=3


def createGraph():
	"create tensors"
	tinputs=tf.placeholder(tf.float32, [numTimesteps, batchSize, numClasses])
	tseqLen=tf.placeholder(tf.int32, [None]) # list of sequence length in batch
	tbeam=tf.nn.ctc_beam_search_decoder(tinputs, tseqLen, beam_width=1, merge_repeated=False)
	tbest=tf.nn.ctc_greedy_decoder(tinputs, tseqLen, merge_repeated=True)

	return (tinputs, tseqLen, tbeam, tbest)


def getData():
	"get data matrix of size TxBxC with T=3, B=1 and C=2 (one label + blank)"
	seqLen=[numTimesteps]
	inputs=np.asarray([ [[0.51, 0.49]], [[0.49, 0.51]], [[0.51, 0.49]] ], np.float32)
	return (inputs, seqLen) 


def toLabelString(decoderOutput):
	"map sparse tensor from decoder to label string"
	decoded=decoderOutput[0][0]
	idxDict={b:[] for b in range(batchSize)}
	encodedLabels=[[] for i in range(batchSize)]
	for (idxVal, idx2d) in enumerate(decoded.indices):
		value=decoded.values[idxVal]
		batch=idx2d[0]
		encodedLabels[batch].append(value)

	return encodedLabels[0]

def main():
	# initialize
	(tinputs, tseqLen, tbeam, tbest)=createGraph()
	sess=tf.Session()
	sess.run(tf.global_variables_initializer())

	# compute decoded result
	(inputs, seqLen)=getData()
	[retBeam, retBest]=sess.run([tbeam, tbest], {tinputs:inputs, tseqLen:seqLen } )
	print('Beam Search Decoding :', toLabelString(retBeam))
	print('Best Path Decoding   :', toLabelString(retBest))


if __name__ == '__main__':
	main()
```

Output:
```
Beam Search Decoding : [0]
Best Path Decoding   : [0, 0]
```

---

Have I written custom code: code see above
OS Platform and Distribution: N/A
TensorFlow installed from: N/A
TensorFlow version: tested with 1.3 and 1.6
Bazel version: N/A
CUDA/cuDNN version: N/A
GPU model and memory: N/A
Exact command to reproduce: code see above, doc is linked
Mobile device: N/A

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by githubharald** [comment]
YES

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 104 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 119 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by githubharald** [comment]
After half a year of inactivity of this issue, I give up, just leave the error in your documentation, I don't care :-)

**Comment by noahzhy** [comment]
I argee with @githubharald that the result of the greedy and beam decoder is different if beam width and top paths is set to 1. I do not have a strong theoretical background but I found the same problem when I optimizing the model, change with the two different decoders.

**Comment by frgfm** [comment]
After some time spent trying to obtain the same result with the greedy decoder, I ended up finding this issue. This is rather simple: either the documentation needs to be more specific explaining how to obtain the same outputs, which doesn't seem to be possible, or we simply remove this wrong Note to avoid confusion.

Can someone from the TF core team take a look at this? :pray:  @tatianashp (not sure exactly who should be tagged, trying to ping regular maintainers @mihaimaruseac @yongtang @hawkinsp)

This is an easy bug to fix, I'm happy to open a PR if that solves the issue :+1:

**Comment by mihaimaruseac** [comment]
Unfortunately I'm now working on this specific area but PRs are always welcome

**Comment by frgfm** [comment]
> Unfortunately I'm now working on this specific area but PRs are always welcome

@mihaimaruseac I just opened https://github.com/tensorflow/tensorflow/pull/53269 to address this :+1:
